File: js/src/json-schema-dsl/schemaProxy.js
const schemaGetter = require('./schemaGetter');
const schemaSetter = require('./schemaSetter');

function createProxy(target, schemaObject, path = []) {
    return new Proxy(target, new SchemaProxy(schemaObject, path));
}

class SchemaProxy {
    constructor(schemaObject, path) {
        this.$schemaObject = schemaObject;
        this.$path = path;
    }

    get(target, fieldName, receiver) {
        if (fieldName === '$get' || fieldName === '$set') {
            return this[fieldName](target);
        }
        return createProxy(target, this.$schemaObject, this.$path.concat([fieldName]));
    }

    $get(target) {
        return () => schemaGetter(this.$schemaObject, this.$path, target);
    }

    $set(target) {
        return value => schemaSetter(this.$schemaObject, this.$path, target, value);
    }
}

module.exports = {
    createProxy: createProxy
};

File: python/src/schema_proxy.py
import schema_getter
import schema_setter


class Proxy:
    def __init__(self, target, schema_object, path=None):
        self._TARGET_ = target
        self._SCHEMA_OBJECT_ = schema_object
        self._PATH_ = path or []

    def __getattr__(self, name):
        return Proxy(self._TARGET_, self._SCHEMA_OBJECT_, self._PATH_ + [name])

    def __getitem__(self, key):
        return Proxy(self._TARGET_, self._SCHEMA_OBJECT_, self._PATH_ + [key])

    def _GET_(self):
        return schema_getter.getter(self._SCHEMA_OBJECT_, self._PATH_, self._TARGET_)

    def _SET_(self, value):
        schema_setter.setter(self._SCHEMA_OBJECT_, self._PATH_, self._TARGET_, value)

